src/stdlib/RandomMeasures.jl

```diff
@@ -128,41 +128,31 @@ function distribution(d::SizeBiasedSamplingProcess{<:DirichletProcess})
     return LocationScale(zero(α), d.surplus, Beta(one(α), α))
 end
 
-function _logpdf_table(d::DirichletProcess, m::AbstractVector{Int})
+function _logpdf_table(d::DirichletProcess{T}, m::AbstractVector{Int}) where {T<:AbstractFloat}
     # compute the sum of all cluster counts
     sum_m = sum(m)
 
-    # shortcut if all cluster counts are zero
-    dα = d.α
-    T = typeof(dα)
-    iszero(sum_m) && return zeros(T, 1)
-
     # pre-calculations
-    z = log(sum_m - 1 + dα)
+    z = log(sum_m - 1 + d.α)
 
     # construct the table
-    K = length(m)
-    table = Vector{T}(undef, K)
-    contains_zero = false
-    @inbounds for i in 1:K
-        mi = m[i]
-
-        if iszero(mi)
-            if contains_zero
-                table[i] = -Inf
-            else
-                table[i] = log(dα) - z
-                contains_zero = true
-            end
-        else
-            table[i] = log(mi) - z
-        end
+    has_zeros = any(i -> i == 0, m)
+    K = has_zeros ? length(m) : length(m)+1
+    table = ones(T, K) * T(-Inf)
+
+    if iszero(sum_m)
+        table[1] = T(0)
+        return table
     end
 
-    if !contains_zero
-        push!(table, log(dα) - z)
+    contains_zero = false
+    @inbounds for i in (has_zeros ? findall(j -> j>0, m) : 1:(K-1))
```

Comment:
The creation of a new array by `findall` could be changing this to
```suggestion
    @inbounds for i in 1:length(m)
        m[i] > 0 || continue
```

Reply:
I'll time what is faster. Thanks.

Reply:
This turns out to be significantly slower. So I'll keep the current version.

Reply:
Ah interesting. Maybe that's due to most of the entries being positive actually? One could both avoid the unneeded checks and the use of `findall` by instead using
```julia
if first_zero === nothing
    @inbounds for i in 1:length(m)
        table[i] = ....
    end
else
    @inbounds for i in 1:length(m)
        mi = m[i]
        mi > 0 && (table[i] = ....)
    end
end
```
That should not lead to any performance regressions for `first_zero === nothing` and I would assume that is faster for `first_zero !== nothing` (since `findall` has to perform those checks as well and additionally creates an array).

Reply:
This seems to depend a lot on the number of zero entries. If it's 50/50, findall performs better than your proposal while it, of course, slows down in case we have very few zeros.

Reply:
This feels like over optimising stuff.

Reply:
> This seems to depend a lot on the number of zero entries. If it's 50/50, findall performs better than your proposal while it, of course, slows down in case we have very few zeros.

I'm confused - as expected, in my benchmarks `findall` is always significantly slower than just checking each entry in a for loop, regardless of the number of zeros:
```julia
julia> using BenchmarkTools

julia> function f_findall!(out, x)
          @inbounds for i in findall(xi -> xi > 0, x)
              out[i] = log(x[i])
           end
           out
       end

julia> function f_if!(out, x)
          @inbounds for i in 1:length(x)
              xi = x[i]
              xi > 0 && (out[i] = log(xi))
           end
           out
       end
julia> out = Vector{Float64}(undef, 10);

julia> x = vcat(zeros(5), rand(5));

julia> @btime f_findall!($out, $x);
  194.050 ns (7 allocations: 352 bytes)

julia> @btime f_if!($out, $x);
  56.534 ns (0 allocations: 0 bytes)

julia> y = rand(10);

julia> @btime f_findall!($out, $y);
  275.958 ns (8 allocations: 496 bytes)

julia> @btime f_if!($out, $y);
  90.970 ns (0 allocations: 0 bytes)

julia> z = vcat(0, rand(9));

julia> @btime f_findall!($out, $z);
  267.887 ns (8 allocations: 496 bytes)

julia> @btime f_if!($out, $z);
  83.254 ns (0 allocations: 0 bytes)

julia> a = zeros(10);

julia> @btime f_findall!($out, $a);
  60.132 ns (4 allocations: 144 bytes)

julia> @btime f_if!($out, $a);
  16.722 ns (0 allocations: 0 bytes)

julia> out_large = Vector{Float64}(undef, 100);

julia> x_large = vcat(zeros(50), rand(50));

julia> @btime f_findall!($out_large, $x_large);
  910.286 ns (10 allocations: 1.28 KiB)

julia> @btime f_if!($out_large, $x_large);
  412.361 ns (0 allocations: 0 bytes)

julia> y_large = rand(100);

julia> @btime f_findall!($out_large, $y_large);
  1.677 μs (11 allocations: 2.34 KiB)

julia> @btime f_if!($out_large, $y_large);
  746.392 ns (0 allocations: 0 bytes)

julia> z_large = vcat(0, rand(99));

julia> @btime f_findall!($out_large, $z_large);
  1.829 μs (11 allocations: 2.34 KiB)

julia> @btime f_if!($out_large, $z_large);
  745.020 ns (0 allocations: 0 bytes)

julia> a_large = zeros(100);

julia> @btime f_findall!($out_large, $a_large);
  132.950 ns (4 allocations: 144 bytes)

julia> @btime f_if!($out_large, $a_large);
  81.840 ns (0 allocations: 0 bytes)
```
So to me these benchmarks seem to indicate that one shouldn't use `findall` in this case, since the alternative is significantly faster.